devdoc/worker_thread_requirements.md

```diff
@@ -125,21 +119,20 @@ MOCKABLE_FUNCTION(, WORKER_THREAD_SCHEDULE_PROCESS_RESULT, worker_thread_schedul
 
 **SRS_WORKER_THREAD_01_042: [** If `sm_exec_begin` does not grant the execution, `worker_thread_schedule_process` shall fail and return `WORKER_THREAD_SCHEDULE_PROCESS_INVALID_STATE`. **]**
 
-**SRS_WORKER_THREAD_01_017: [** `worker_thread_schedule_process` shall set the process event created in `worker_thread_create`. **]**
+**SRS_WORKER_THREAD_01_017: [** `worker_thread_schedule_process` shall set the state to `WORKER_THREAD_STATE_PROCESS_ITEM`. **]**
 
 **SRS_WORKER_THREAD_01_043: [** `worker_thread_schedule_process` shall call `sm_exec_end`. **]**
 
 **SRS_WORKER_THREAD_01_015: [** On success `worker_thread_schedule_process` shall return `WORKER_THREAD_SCHEDULE_PROCESS_OK`. **]**
 
-**SRS_WORKER_THREAD_01_018: [** If any other error occurs, `worker_thread_schedule_process` shall fail and return `WORKER_THREAD_SCHEDULE_PROCESS_ERROR`. **]**
-
 ### worker_thread
 
-**SRS_WORKER_THREAD_01_019: [** The worker thread started by `worker_thread_create` shall wait for the 2 events: thread shutdown event and execute worker function event. **]**
+**SRS_WORKER_THREAD_01_019: [** The worker thread started by `worker_thread_create` shall get the module state. **]**
 
-**SRS_WORKER_THREAD_01_020: [** When the shutdown event is signaled, the worker thread shall exit. **]**
+**SRS_WORKER_THREAD_01_020: [** If the module state is `WORKER_THREAD_STATE_CLOSE`, the worker thread shall exit. **]**
 
-**SRS_WORKER_THREAD_01_021: [** When the execute worker function event is signaled, the worker thread shall call the `worker_func` function passed to `worker_thread_create` and it shall pass `worker_func_context` as argument. **]**
+**SRS_WORKER_THREAD_01_021: [** If the module state is `WORKER_THREAD_STATE_PROCESS_ITEM`, the worker thread shall call the `worker_func` function passed to `worker_thread_create` and it shall pass `worker_func_context` as argument... **]**
 
-**SRS_WORKER_THREAD_01_025: [** In case of any error, the worker thread shall exit. **]**
+**SRS_WORKER_THREAD_11_001: [** ... and set the module state to idle if it has not been changed. **]**
 
+**SRS_WORKER_THREAD_11_002: [** If the module state is `WORKER_THREAD_STATE_IDLE`, the worker thread shall wait for the state to transition to something else. **]**
```

Comment:
>**]** [](http://example.com/codeflow?start=159&length=5)

having read everything until now, I find that the promises of this module are a bit lacking in regards to timing and GREAT care needs to be exercised by the user.

Here's what I mean. Let's say the user wants to execute 3 (times) the `worker_func`. Let's say `worker func` take 100 ms to execute.

Here's 2 timelines
Timeline1:
T+0 => schedule process
T+1 => `worker func` starts to execute.
T+101 = `worker func` is done executing(1st execution).

T+200 => schedule process
T+201 => `worker func` starts to execute.
T+301 = `worker func` is done executing(2nd execution).

T+400 => schedule process
T+401 => `worker func` starts to execute.
T+501 = `worker func` is done executing(3rd execution).

Timeline2:
T+0 => schedule process
T+1 => `worker func` starts to execute.
T+25 => schedule process (sets state to "WORKER_THREAD_STATE_PROCESS_ITEM as per `SRS_WORKER_THREAD_01_017`)
T+75 => schedule process (sets state to "WORKER_THREAD_STATE_PROCESS_ITEM as per `SRS_WORKER_THREAD_01_017`)
T+101 => `worker func` is done executing(1st execution)
T+102 => `worker func` executes again (because state is WORKER_THREAD_STATE_PROCESS_ITEM)
T+202 => `worker func` ends and NOTHING executed anymore.

The difference between Timeline 1 and Timeline 2 is that in Timeline 2 there are 2 executions, while in Timeline 1 there are 3, and the only difference is in the timing when `schedule process` is called.

I believe the user's intention is to always have 3 executions. And we can talk about it.

Reply:
I too was concerned about this when I looked at it.  It looks like the contract is that schedule process must not be called until WORKER_FUNC has been called (this is stated on cryptic line 13 in the spec).  If this is indeed the contract (which I looked at the ebs code that is calling this module) then the 2nd timeline breaks the contract.

Reply:
This module needs to prohibit the overcluttering of calls (which results in missed executions).